File: backend/src/main/java/com/seatlock/common/error/DomainException.java
package com.seatlock.common.error;

import java.util.Map;
import lombok.Getter;

/**
 * 도메인 에러의 표준형. GlobalExceptionHandler가 { code, message, details? }로 직렬화한다.
 * details는 클라이언트가 후속 행동을 결정하는 데 필요한 최소 정보만 담는다
 * (예: SEAT_ALREADY_TAKEN의 충돌 좌석 목록).
 */
@Getter
public class DomainException extends RuntimeException {

    private final ErrorCode errorCode;
    private final transient Map<String, Object> details;

    public DomainException(ErrorCode errorCode) {
        this(errorCode, null);
    }

    public DomainException(ErrorCode errorCode, Map<String, Object> details) {
        super(errorCode.getMessage());
        this.errorCode = errorCode;
        this.details = details;
    }
}

File: backend-nest/src/common/errors/domain.exception.ts
import { HttpException } from '@nestjs/common';

/**
 * 도메인 에러의 표준형.
 * HTTP 상태코드만으로는 "왜 실패했는가"를 구분할 수 없으므로(409가 좌석 선점 실패인지
 * 이메일 중복인지), 기계가 분기할 수 있는 안정적인 code를 계약으로 제공한다.
 */
export class DomainException extends HttpException {
  constructor(
    status: number,
    public readonly code: string,
    message: string,
    public readonly details?: Record<string, unknown>,
  ) {
    super({ code, message, ...(details !== undefined ? { details } : {}) }, status);
  }
}
